### PR TITLE
feat(testing): support for Jest 27

### LINKED ISSUE #5771: Support Jest 27
## Description

Jest 27 has just released and it only supports Node.js LTS and v15.
https://jestjs.io/blog/2021/05/25/jest-27
https://github.com/facebook/jest/releases/tag/v27.0.0

## Motivation

We can use the latest version of Jest.

## Suggested Implementation

A migration for `"testEnvironment": "jsdom"` may be needed.

### PR DESCRIPTION
**Pull request by AgentEnder**
- Sets default testEnvironment to web
- Updates node generators to specify testEnvironment: node
- Provides migrations to update existing jest.config.js files
- Bump angular-preset-jest to latest

Fixes #5771

### CODE REVIEW

**Comment by Coly010** [comment]
With the bump to `jest-preset-angular@9.0.1` could you do me a favour and check the following test case for errors:

---
## Test Case 

### Steps

 - Create Empty Nx Workspace using npm (`npx create-nx-workspace@999.9.9` (999.9.9 is the version number that is likely in the local registry if you're following the steps in `CONTRIBUTING.md`))
 - Install `@nrwl/angular` as devDep
 - Generate an angular application
 - Run the tests using `npm run nx test` (specifically `npm run` so that it uses the local version in the repository

### Expected Result
There should be no errors and the tests should run and complete successfully

--- 

The reason for this is related to npm package resolution and package hoisting. We need to make sure that by turning off the `legacy-peer-deps` flag, we still get the correct behaviour.

**Comment by AgentEnder** [comment]
@Coly010 running that test manually the test suite runs without issue.

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@nrwl/nx-source",
-  "version": "12.3.5",
+  "version": "99.123.1",
```
Comment (juristr) [inline package.json:3]: I think this got in accidentally 🙂

**Review by juristr — DISMISSED** [review]

File: package.json (hunk 1 above)
Comment (AgentEnder) [reply package.json:3]: Thanks, my bad.

File: packages/jest/src/generators/jest-project/files/jest.config.js__tmpl__ [hunk 2]
```diff
@@ -21,6 +16,9 @@ module.exports = {
     moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],<% } else { %>
     moduleFileExtensions: ['ts', 'js', 'html'],<% } %><% } %>
   coverageDirectory: '<%= offsetFromRoot %>coverage/<%= projectRoot %>'<% if(!skipSerializers) { %>,
+  transform: {
+    '^.+\\.(ts|js|html)$': 'jest-preset-angular'
+  },
```
Comment (Cammisuli) [inline packages/jest/src/generators/jest-project/files/jest.config.js__tmpl__:21]: we'd have to wrap this logic in a same type of if as it was before:

```suggestion
<%if (setupFile === 'angular') { %>
   transform: {
     '^.+\\.(ts|js|html)$': 'jest-preset-angular'
   },
<% } %>
```

Also watch out for the previous `if` (`<% if(!skipSerializers) {`). This whole template is tricky so make sure to double check all the options when testing

File: packages/node/src/generators/application/application.spec.ts [hunk 3]
```diff
@@ -330,6 +330,7 @@ describe('app', () => {
         "module.exports = {
           displayName: 'my-node-app',
           preset: '../../jest.preset.js',
+          testEnvironment: 'node',
```
Comment (Cammisuli) [inline packages/node/src/generators/application/application.spec.ts:333]: We need to set the `testEnvironment` for node libs as well.

File: packages/jest/migrations.json [hunk 4]
```diff
@@ -150,6 +162,19 @@
...
+      "version": "12.4.0-beta.1",
+      "packages": {
+        "jest": {
+          "version": "27.0.3",
+          "alwaysAddToPackageJson": false
+        },
+        "ts-jest": {
+          "version": "27.0.2",
+          "alwaysAddToPackageJson": false
+        }
```
Comment (Cammisuli) [inline packages/jest/migrations.json:176]: I think we need to include `jest-preset-angular` here as well

**Review by Cammisuli — CHANGES_REQUESTED** [review]
Just a couple things that need to change, but otherwise, things look good! Thank you!

File: packages/jest/src/generators/jest-project/files/jest.config.js__tmpl__ (hunk 2 above)
Comment (AgentEnder) [reply packages/jest/src/generators/jest-project/files/jest.config.js__tmpl__]: Resolved by splitting out the angular templates into files-angular.

**Comment by ahnpnl** [comment]
Pls pin `ts-jest@27.0.3` and `jest-preset-angular@9.0.2` to have all Node 12, 14, 15, 16 support

**Comment by ahnpnl** [comment]
Sorry for jump in again, `jest-preset-angular@9.0.3` should be used. This release only contains a fix to correct `peerDependencies`.

 The bug most likely won’t happen to angular apps inside Nrwl nx but better stay on safe side.

**Comment by AgentEnder** [comment]
No worries @ahnpnl , I'll get that bumped up.

**Comment by mehrad-rafigh** [comment]
@juristr @Cammisuli When can this be merged and released?

**Comment by Cammisuli** [comment]
@mehrad-rafigh I'm going to do a sanity check on Friday to make sure everything is golden 👌🏼

**Review by Cammisuli — APPROVED** [review]
Things look good from my side:
* `create-nx-workspace` with angular, react, node, nest, express have the proper configs and tests run fine
* migrating multiple repos with the migrations built in this PR are running correctly, and configs are updated.

**Pull request merged**